stdenvAdapters: add useMoldLinker

Artturin opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
bintools-wrapper: wrap all 'ld.*'

allows using wrapBintoolsWith with all linkers

```
$ nix build ".#binutils"
$ ls ./result/bin/ld*
./result/bin/ld*  ./result/bin/ld.bfd*  ./result/bin/ld.gold*

$ nix build "nixpkgs#binutils"
$ ls ./result/bin/ld*
./result/bin/ld*  ./result/bin/ld.bfd* ./result/bin/ld.gold*
```

Ericson2314 commented on pkgs/build-support/bintools-wrapper/default.nix:
```diff
@@ -192,7 +192,7 @@ stdenv.mkDerivation {
     '') + ''
 
-      for variant in ld.gold ld.bfd ld.lld; do
+      for variant in ld.gold ld.bfd ld.lld ld.mold; do
```
```suggestion
      for variant in ld.*; do
```
Is it time to do this instead?

VitalyAnkh replied:
It's time to do so! Doing this also makes convenience for more new linkers.

Ericson2314 replied:
OK then apply that that diff / squash history as you see fit!

Artturin commented:
binutils works
```
$ nix build ".#binutils"
$ ls ./result/bin/ld*
./result/bin/ld*  ./result/bin/ld.bfd*  ./result/bin/ld.gold*

$ nix build "nixpkgs#binutils"
$ ls ./result/bin/ld*
./result/bin/ld*  ./result/bin/ld.bfd* ./result/bin/ld.gold*

$ ./result/bin/ld
/nix/store/sc0h1nnb9rxbiknq692dq9n3wwsrhbbl-binutils-2.40/bin/ld: no input files
$ ./result/bin/ld.bfd
/nix/store/sc0h1nnb9rxbiknq692dq9n3wwsrhbbl-binutils-2.40/bin/ld.bfd: no input files
$ ./result/bin/ld.gold
/nix/store/sc0h1nnb9rxbiknq692dq9n3wwsrhbbl-binutils-2.40/bin/ld.gold: fatal error: no input files
```


```
$ nix build --impure --expr "with import ./. {}; wrapBintoolsWith { bintools = mold; }"
$ ls ./result/bin
ld*  ld.mold*
$ ./result/bin/ld.mold
mold: fatal: -m option is missing
$ ./result/bin/ld
./result/bin/ld: line 254: /nix/store/nkxfnbmkpv53y7fibkxq4c9aq21ik606-mold-1.10.1/bin/ld: No such file or directory
```

https://github.com/NixOS/nixpkgs/blob/63b5955814db30d2e2ff7157aaa5665b502ed2f4/pkgs/build-support/bintools-wrapper/default.nix#L186

Ericson2314 commented:
So I guess there are two issues:

1. Mold doesn't install an `ld`
2. Mold doesn't support all the flags our wrapper uses yet

Right?

Artturin commented:
They install a ld in libexec

https://github.com/rui314/mold/blob/main/CMakeLists.txt#L453

```
$ tree result
result
├── bin
│   ├── ld.mold -> mold
│   └── mold
├── lib
│   └── mold
│       └── mold-wrapper.so
├── libexec
│   └── mold
│       └── ld -> ../../bin/mold
└── share
    ├── doc
    │   └── mold
    │       └── LICENSE
    └── man
        └── man1
            ├── ld.mold.1.gz -> mold.1.gz
            └── mold.1.gz
```

Ericson2314 commented:
That sounds good. The wrapper could also promte one of the `ld.*` if there is no `ld`.

azahi commented:
> Mold doesn't install an ld

I guess we can just apply this [patch](https://gist.github.com/azahi/184baff55b7dbf9d230000e3ba74f699) in the scope of this PR to avoid opening more PRs.
```diff
diff --git i/pkgs/development/tools/mold/default.nix w/pkgs/development/tools/mold/default.nix
index bb55ba80796..4b2b35fb26f 100644
--- i/pkgs/development/tools/mold/default.nix
+++ w/pkgs/development/tools/mold/default.nix
@@ -1,6 +1,7 @@
 { lib
 , stdenv
 , fetchFromGitHub
+, makeWrapper
 , cmake
 , mimalloc
 , ninja
@@ -45,6 +46,10 @@ stdenv.mkDerivation rec {
     "-faligned-allocation"
   ];
 
+  postInstall = ''
+    makeWrapper "$out/bin/mold" "$out/bin/ld"
+  '';
+
   passthru.tests.version = testers.testVersion { package = mold; };
 
   meta = with lib; {
```

> Mold doesn't support all the flags our wrapper uses yet

I'm not sure about a 100% flag parity to GNU `ld`, but I've used `mold` as a drop-in replacement for it in a few relatively big projects for a while now (not on NixOS) and had alsmost no issues.

Ericson2314 commented:
The fact that they have an `ld.mold` indicates to me they at least *intend* to hit feature parity. That is reassuring ;).

Artturin committed:
bintools-wrapper: dont wrap ld if it doesn't exist

not all linkers have a ld binary in bin

also note the '${ld:-}' which allows users to set the ld path with a env
var

> '${foo:-val}' $foo, or val if unset (or null)

Artturin committed:
stdenvAdapters: add useMoldLinker

Artturin commented:
i removed `mold: symlink ld to ld.mold for compat with bintools-wrapper` because its not necessary now

this works and the resulting binaries have `mold 1.10.1 (compatible with GNU ld)`

```diff
diff --git a/pkgs/tools/misc/man-db/default.nix b/pkgs/tools/misc/man-db/default.nix
index 2bdd49f672f..8976d3314cb 100644
--- a/pkgs/tools/misc/man-db/default.nix
+++ b/pkgs/tools/misc/man-db/default.nix
@@ -12,9 +12,14 @@
 , stdenv
 , zstd
 , autoreconfHook
+, stdenvAdapters
 }:
 
-stdenv.mkDerivation rec {
+let
+  stdenv' = stdenvAdapters.useMoldLinker stdenv;
+in
+
+stdenv'.mkDerivation rec {
   pname = "man-db";
   version = "2.11.1";
```
